compiler/src/dmd/globals.d

```diff
@@ -350,21 +353,51 @@ extern (C++) struct Global
 
     extern (C++) void _init()
     {
-        global.errorSink = new ErrorSinkCompiler;
+        errorSink = new ErrorSinkCompiler;
 
         this.fileManager = new FileManager();
         version (MARS)
         {
-            vendor = "Digital Mars D";
+            compileEnv.vendor = "Digital Mars D";
 
             // -color=auto is the default value
             import dmd.console : detectTerminal;
             params.color = detectTerminal();
         }
         else version (IN_GCC)
         {
-            vendor = "GNU D";
+            compileEnv.vendor = "GNU D";
+        }
+        compileEnv.versionNumber = parseVersionNumber(_version);
+
+        /* Initialize date, time, and timestamp
+         */
+        import core.stdc.time;
+        import core.stdc.stdlib : getenv;
+
+        time_t ct;
+        // https://issues.dlang.org/show_bug.cgi?id=20444
+        if (auto p = getenv("SOURCE_DATE_EPOCH"))
+        {
+            if (!ct.parseDigits(p[0 .. strlen(p)]))
+                errorSink.error(Loc.initial, "value of environment variable `SOURCE_DATE_EPOCH` should be a valid UNIX timestamp, not: `%s`", p);
         }
+        else
+            core.stdc.time.time(&ct);
+        const p = ctime(&ct);
+        assert(p);
+
+        __gshared char[11 + 1] date = 0;        // put in BSS segment
+        __gshared char[8  + 1] time = 0;
+        __gshared char[24 + 1] timestamp = 0;
```

Comment:
Don't need to be `__gshared`

Reply:
They do because as locals they will disappear when the function exits.

Reply:
I think `dup`ing them is cleaner than keeping a `__gshared` in a function.